pkg/networkservice/common/updatepath/common.go

```diff
@@ -17,74 +17,76 @@
+		2.1 if path has next segment available and next name is segmentName, take Id from next path segment
+		2.2 if no next path segment available, it will add one more path segment and generate new Id, update connection.Id
+		2.3 if path has next segment available, but next name is not equal to segmentName, will return error.
+	3. if Index == 0, and there is no current segment will add one.
+*/
+func updatePath(conn *networkservice.Connection, segmentName string) (*networkservice.Connection, error) {
+	if conn.Path == nil {
+		conn.Path = &networkservice.Path{}
 	}
-
 	// If we don't have a Path, add one
-	if conn.GetPath() == nil {
-		conn.Path = &networkservice.Path{}
+	if len(conn.GetPath().GetPathSegments()) == 0 {
+		conn.Path.Index = 0
+		if conn.Id == "" {
+			// Generate new ID for connection and segment.
+			conn.Id = uuid.New().String()
+		}
+		// Add current segment to list
+		conn.Path.PathSegments = append(conn.Path.PathSegments, &networkservice.PathSegment{
+			Name: segmentName,
+			Id:   conn.Id,
+		})
+		return conn, nil
 	}
 
 	path := conn.GetPath()
 
-	// Make sure index isn't out of bound
-	if (len(path.GetPathSegments()) > 0) && int(path.GetIndex()) >= len(path.GetPathSegments()) {
-		return errors.Errorf("NetworkServiceRequest.Connection.Path.Index(%d) >= len(NetworkServiceRequest.Connection.Path.PathSegments)(%d)",
-			path.GetIndex(), len(path.GetPathSegments()))
+	if int(path.Index) < len(path.PathSegments) && path.PathSegments[path.Index].Name == segmentName {
+		// We already in current segment, just update connection Id, no need to increment
+		conn.Id = path.PathSegments[path.Index].Id
+		return conn, nil
 	}
 
-	// If this isn't already one of our tokens for *this* connectionId ... then update the next one
-	if (len(path.GetPathSegments()) > 0) && !(path.GetPathSegments()[path.GetIndex()].GetName() == u.name &&
-		path.GetPathSegments()[path.GetIndex()].GetId() == conn.GetId()) {
-		path.Index++
-	}
+	// We need to move to next item
+	nextIndex := int(path.Index) + 1
 
-	// If we don't already have a PathSegment at the new index... append an empty one
-	if int(path.GetIndex()) >= len(path.GetPathSegments()) {
-		path.PathSegments = append(path.PathSegments, &networkservice.PathSegment{})
+	if nextIndex > len(path.GetPathSegments()) {
+		// We have index > segments count
+		return nil, errors.Errorf("NetworkServiceRequest.Connection.Path.Index+1==%d should be less or equal len(NetworkServiceRequest.Connection.Path.PathSegement)==%d",
+			nextIndex, len(path.GetPathSegments()))
 	}
-
-	// Extract the authInfo:
-	var authInfo credentials.AuthInfo
-	if p, exists := peer.FromContext(ctx); exists {
-		authInfo = p.AuthInfo
+	if nextIndex < len(path.GetPathSegments()) && path.GetPathSegments()[nextIndex].Name != segmentName {
+		// We have next, but name is different, so let's update Path segment in this case.
+		// Id should be same, since next chain elements may depend on it.
```

Comment:
I don't quite follow why the id should be the same in this case.  As I understand it this is the case where we have:

Say we have a chain:

A(id:1) B(id:2) C(id:3) D(id:4)

B goes away, and A heals to G, G sees the entry for B(id:2) ... why should G reuse id:2 ?

Reply:
Because with id:2 it will go to C, and C could have a map of id:2 to some internal structures, and if it will be different, new connection will be established instead, and previous one will be closed by timeout.

Reply:
C shouldn't be mapping B's id:2 to an internal structure... it should be tracking any of it state data with its own id:3 for the connection.

Reply:
In this case it should be fine to generate new ID :-)